### ISSUE TITLE
Support custom serializer and transformer

### ISSUE THREAD

**Issue by xbrancher** [sneas/ionic-native-http-connection-backend#154]
### Subject of the issue
I need to use `urlencoded` serializor and `bodyToObject` transformer with `application/x-www-form-urlencoded` header.

### Environment
* version of angular is 12
* version of ionic-native-http-connection-backend is 9
* version of @ionic/angular is 5
* version of @ionic-native/http is 5

### Steps to reproduce
Set `application/x-www-form-urlencoded` header and pass HttpParams as request body. 

#### Real case scenario:
I am using `angular-oauth2-oidc` with `keycloak` identity server and fetcing token with password flow raise an error: `415 undefined","error":{"error":"RESTEASY003065: Cannot consume content type"}`

### Expected behaviour
Expected serializerAndData should be `urlencoded` and transfored data wiith `bodyToObject`.

### Actual behaviour
For now `multipart` serializor is used.

**Comment by sneas** [comment]
Hi. Thank you for raising the issue and for providing the PR.

The whole point of this library is the transparent usage of HttpClient through the Native HTTP Cordova plugin. In other words, the expected behavior of the plugin must be identical to the behavior of HttpClient without the plugin. The provided PR breaks the transparency by enforcing a specific behavior with the manipulation of the headers. In other words: the PR makes Angular's `HttpClient` work differently with and without `ionic-native-http-connection-backend`. This will impose implications for the future support of `ionic-native-http-connection-backend` due to the lack of standardization of the given approach and exposing the internals of the library.

As far as I understand from the issue, the library performs incorrect requests with a certain combination of the `Content-Type` header and the type of request body. In other words, the `HttpClient` with `ionic-native-http-connection-backend` performs not the same requests to the server as `HttpClient` without the `ionic-native-http-connection-backend`. This issue must be addressed by aligning the overall behavior of the library.

I've just created a simple example to reproduce the issue: https://github.com/sneas/ionic-5-http-test/blob/issue-154/src/app/tab1/tab1.page.ts

You can see it in action on iOS device by running `npm run ios`.

As far as I can see, there is indeed a difference between request body objects:

On device using `Native HTTP` as a `HttpClient` backend:
![image](https://user-images.githubusercontent.com/144651/143764348-b439a84e-8a66-4ce9-a658-b578c402a4b3.png)

In browser without `Native HTTP`:
![image](https://user-images.githubusercontent.com/144651/143764413-9df15386-d14e-41d3-b770-e9834403a43f.png)

The `form` object is different in both cases (string VS object), but they make no sense in both cases, so I don't see it as a problem of posting `application/x-www-form-urlencoded` with `HttpParams` body. There is something else.

I assume that `angular-oauth2-oidc` performs the request differently. With the request body which makes sense at least without `ionic-native-http-connection-backend`. Thus, we need to understand what request is performed correctly without `ionic-native-http-connection-backend` and incorrectly with `ionic-native-http-connection-backend` and fix that. Your help will be extremely appreciated by the library users.

**Comment by xbrancher** [comment]
I agree this may not be the best solution but I think there should be a way to send `application/x-www-form-urlencoded` content type with `urlencoded` serializer. As I see here (https://github.com/silkimen/cordova-plugin-advanced-http)[https://github.com/silkimen/cordova-plugin-advanced-http] default serializer for `application/x-www-form-urlencoded` is `urlencoded` but `ionic-native-http-connection-backend` always use `multipart` serializer. Maybe we should take a look at request body content. We should use `multipart` serializer only if content is multi part not just FormData or HttpParams with `application/x-www-form-urlencoded` header.

**Comment by sneas** [comment]
Hi,

> I think multipart serializer should be used when there is at least one Blob or ArrayBuffer or File inside HttpParams or FormData.

Someone needs to prove that before applying the fix. Don't get me wrong. This library is been used by multiple clients and its behavior has been shaped by multiple iterations. The only valid case for the library change is the deviation between the XHR and `ionic-native-http-connection-backend` request. For now, I can't see any deviations. All my tests do not show major differences between XHR and `ionic-native-http-connection-backend` requests. I know this library is far from perfect and there might be some misalignments. And one of them is, most likely, is the root of your problem. But this must be proved before applying the fix. I can understand this might be annoying for you because you clearly see the problem with the library. And I truly want to help you. Unfortunately, I'm quite busy with other projects and I can't spend more than few hours per month to support the library. That's why I encourage you to use this project https://github.com/sneas/ionic-5-http-test/ and create a valid case of misalignment between XHR and `ionic-native-http-connection-backend` requests. Again, I do want to help you with resolving your problem, but I don't have much time to verify all the theories.

https://github.com/sneas/ionic-5-http-test uses https://httpbin.org to verify requests. Httpbin responds with the request data so we can clearly see the deviations between XHR and `ionic-native-http-connection-backend`. Do to so you can run `npm run ios`. This will create a development server (http://localhost:8100) and open the app in iOS simulator.  You can open http://localhost:8101 in a browser and see XHR requests. At the same time, you can see how identical requests are performed on a device with `ionic-native-http-connection-backend` enabled (you'll have to open the Safari development console for the particular device).

**Comment by xbrancher** [comment]
I added some examples to https://github.com/sneas/ionic-5-http-test/pull/18

As you can see there is difference in `Content-Type` header for http request in browser and ios device:

IOS DEVICE:
![IMG_0040](https://user-images.githubusercontent.com/41447192/143914158-cc7c0778-8b50-49c2-b97a-5db684e496b6.PNG)

BROWSER:

<img width="818" alt="Screenshot 2021-11-29 at 18 05 06" src="https://user-images.githubusercontent.com/41447192/143914728-73218445-31dd-4ae2-a683-cc7fd958817e.png">

**Comment by sneas** [comment]
Thank you for your PR, @xbrancher ! You've made a great job.

Looking at the `HttpParams` interface, I see that it can't contain anything but primitive values, thus we can safely remove this bit https://github.com/sneas/ionic-native-http-connection-backend/blob/master/src/utils/detect-serializer-and-data.ts#L35-L43  (together with `httpParamsToFormData` function) and it will fall back to the last return of the function: https://github.com/sneas/ionic-native-http-connection-backend/blob/master/src/utils/detect-serializer-and-data.ts#L59-L62

I guess I'll need to conduct a better investigation before doing so.

I'm now wondering, is the incorrect `HttpParams` encoding is the source of your problem or there is something else, like incorrect `FormData` encoding?

**Comment by xbrancher** [comment]
Yes I think the combination of `HttpParams` and `application/x-www-form-urlencoded` header is the only problem that library `angular-oauth2-oidc` doesn't work properly.

**Comment by sneas** [comment]
I've just experimented with our idea and created this PR as a result https://github.com/sneas/ionic-native-http-connection-backend/pull/156

The problem arises when HttpParams contains composite values.

XHR sends them as a form value, which looks like:

```json
{
  "param1": ["value1", "value2"]
}
```

And `urlencoded` serializer sends them as a form value which looks like:

```json
{
  "param1[]": ["value1", "value2"]
}
```

That's why I decided to use `urlencoded` serializer only for flat `HttpParams`. Basically, it pretty much reflects what you have said previously 👍:

> I think multipart serializer should be used when there is at least one Blob or ArrayBuffer or File inside HttpParams or FormData.

Will this PR resolve your problem?
